pkg/tls/certRenewer.go

```diff
@@ -117,34 +119,58 @@ func (c *CertRenewer) WriteCACertToSecret(caPEM *PemPair, props CertificateProps
 	logger := c.log.WithName("CAcert")
 	name := generateRootCASecretName(props)
 
+	depl, err := c.client.GetResource("", "Deployment", props.Namespace, config.KyvernoDeploymentName)
+
+	deplHash := ""
+	if err == nil {
+		deplHash = fmt.Sprintf("%v", depl.GetUID())
+	}
+
+	var deplHashSec string = "default"
+	var ok, managedByKyverno bool
+
 	secretUnstr, err := c.client.GetResource("", "Secret", props.Namespace, name)
-	if err != nil {
-		secret := &v1.Secret{
-			TypeMeta: metav1.TypeMeta{
-				Kind:       "Secret",
-				APIVersion: "v1",
-			},
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      name,
-				Namespace: props.Namespace,
-				Annotations: map[string]string{
-					SelfSignedAnnotation: "true",
-				},
-				Labels: map[string]string{
-					ManagedByLabel: "kyverno",
-				},
+	if err == nil {
+		if label, ok := secretUnstr.GetLabels()[ManagedByLabel]; ok {
+			managedByKyverno = label == "kyverno"
+		}
+		deplHashSec, ok = secretUnstr.GetAnnotations()[MasterDeploymentUID]
+	}
+
+	secret := &v1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: props.Namespace,
+			Annotations: map[string]string{
+				SelfSignedAnnotation: "true",
+				MasterDeploymentUID:  deplHash,
 			},
-			Data: map[string][]byte{
-				RootCAKey: caPEM.Certificate,
+			Labels: map[string]string{
+				ManagedByLabel: "kyverno",
 			},
-			Type: v1.SecretTypeOpaque,
-		}
+		},
+		Data: map[string][]byte{
+			RootCAKey: caPEM.Certificate,
+		},
+		Type: v1.SecretTypeOpaque,
+	}
 
-		_, err := c.client.CreateResource("", "Secret", props.Namespace, secret, false)
+	if err != nil && strings.HasSuffix(err.Error(), "not found") {
+		_, err = c.client.CreateResource("", "Secret", props.Namespace, secret, false)
 		if err == nil {
 			logger.Info("secret created", "name", name, "namespace", props.Namespace)
 		}
 		return err
+	} else if managedByKyverno && (!ok || deplHashSec != deplHash) {
+		_, err = c.client.UpdateResource("", "Secret", props.Namespace, secret, false)
+		if err == nil {
+			logger.Info("secret updated", "name", name, "namespace", props.Namespace)
+		}
+		return err
 	}
```

Comment:
Missing the last `else` error handling:
```
if err != nil && strings.HasSuffix(err.Error(), "not found") {

} else if managedByKyverno && (!ok || deplHashSec != deplHash) {

} else {

// handle error

}

Reply:
I think `else if err != nil {...}` would be more appropriate because for `err == nil` we have some updates on the Secret a few lines down.

Reply:
Agreed, that's more accurate.